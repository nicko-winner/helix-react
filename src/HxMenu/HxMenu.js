import React from 'react';
import PropTypes from 'prop-types';

/**
 * <HxMenu /> wrapper for <hx-menu />
 *
 * @example docs https://rackerlabs.github.io/helix-ui/components/menus/
 * @example element https://rackerlabs.github.io/helix-ui/elements/hx-menu/
 */

class HxMenu extends React.PureComponent {
  componentDidMount() {
    if (this.hxRef) {
      this.hxRef.addEventListener('close', this.props.onClose);
      this.hxRef.addEventListener('open', this.props.onOpen);
    }
  }

  componentWillUnmount() {
    if (this.hxRef) {
      this.hxRef.removeEventListener('close', this.props.onClose);
      this.hxRef.removeEventListener('open', this.props.onOpen);
    }
  }

  setRef = (element) => { this.hxRef = element; };

  render() {
    const { onOpen, onClose, ...hxProps } = this.props;
    return (
      <hx-menu {...hxProps} ref={this.setRef}>
        {this.props.children}
      </hx-menu>
    );
  }
}

HxMenu.propTypes = {
  onOpen: PropTypes.func,
  onClose: PropTypes.func,
  children: PropTypes.array
};

HxMenu.defaultProps = {
  onOpen: () => null,
  onClose: () => null
};

export default HxMenu;
