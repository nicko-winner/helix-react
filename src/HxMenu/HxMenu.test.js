import * as enzyme from 'enzyme';
import HxMenu from './HxMenu';
import React from 'react';

const mount = (props = {}) => {
  return enzyme.mount(
    <HxMenu id='menuId' {...props}>
      <hx-menuitem>Action 1</hx-menuitem>
      <hx-menuitem>Action 2</hx-menuitem>
      <hx-menuitem>Action 3</hx-menuitem>
    </HxMenu>);
};

describe('HxMenu', () => {
  it('renders correctly with slot/child elements', () => {
    expect(mount().html()).toMatchSnapshot();
  });

  it('props.onOpen is called on WebComponent `open` event', () => {
    const onOpen = jest.fn();
    mount({onOpen})
      .instance()
      .hxRef
      .dispatchEvent(new Event('open'));

    expect(onOpen.mock.calls.length).toBe(1);
  });

  it('props.onClose is called on WebComponent `close` event', () => {
    const onClose = jest.fn();
    mount({onClose})
      .instance()
      .hxRef
      .dispatchEvent(new Event('close'));

    expect(onClose.mock.calls.length).toBe(1);
  });

  it('componentWillUnmount removes all event listeners', () => {
    const instance = mount().instance();
    instance.hxRef.removeEventListener = jest.fn();
    instance.componentWillUnmount();
    const listenersRemoved = instance.hxRef.removeEventListener.mock.calls.map((event) => event[0]);

    expect(listenersRemoved).toEqual(expect.arrayContaining(['open', 'close']));
  });
});
