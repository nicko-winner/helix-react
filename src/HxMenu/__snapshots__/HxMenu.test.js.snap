// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HxMenu renders correctly with slot/child elements 1`] = `"<hx-menu id=\\"menuId\\"><hx-menuitem>Action 1</hx-menuitem><hx-menuitem>Action 2</hx-menuitem><hx-menuitem>Action 3</hx-menuitem></hx-menu>"`;
